### DISCUSSION TITLE
Incompleteness of CVC4 - can we do something about it?

### DISCUSSION THREAD

**Discussion by zvonimir** [cvc5/cvc5#7245] [Q&A] [answered]
We are hitting what seems to be an incompleteness issue with CVC4 when we use it with SMACK on some benchmarks (see https://github.com/smackers/smack/issues/714).

When running CVC4 on the attached query ([test.zip](https://github.com/CVC4/CVC4/files/6251242/test.zip)) using `cvc4 --incremental test.smt2` the output we get is:
```
unknown
(:reason-unknown incomplete)
```

I suspect the reason is incompleteness in the presence of quantifiers. Having said that, given that this is a small query that other solvers can handle, I wonder if we are missing some obvious options for CVC4 that we should select.

The version I am using is the following:
```
This is CVC4 version 1.8 [git HEAD 52479010]
```

Thanks for your help!

**Comment by nafur** [comment]
A minimized input:
```
(declare-fun y (Int) Int)
(assert (forall ((q Int)) (! false :pattern ((y 0)))))
(check-sat)
(get-info :reason-unknown)
```
```
$ bin/cvc4 out.smt2 
unknown
(:reason-unknown incomplete)
```

**Comment by ajreynol** [comment, marked as answer]
Hi Zvonomir,

There are several options that CVC4 can use for quantified formulas that are not enabled by default.  Here are three suggestions:

`--full-saturate-quant`: This enables enumerative instantiation (https://homepage.divms.uiowa.edu/~ajreynol/tacas18.pdf).

`--finite-model-find`: This enables model-based quantifier instantiation (https://homepage.divms.uiowa.edu/~ajreynol/cade24.pdf) which is based on finite model finding.

`--fmf-bound`: This enables finite instantiation based on finite integer bounds.

My suggestion is typically: 
If you think your problem is unsat and you are getting unknown, use `--full-saturate-quant`.
If you think your problem is sat and you are getting unknown, then if your quantifiers are over uninterpreted sorts, use `--finite-model-find`; if your quantifiers are over finite (integer) ranges, use `--fmf-bound`.

For the attached example, adding `--fmf-bound` allows CVC4 to answer "sat" for the last query.  For Gereon's example above, adding `--full-saturate-quant` allows CVC4 to answer "unsat".

It should be noted however that `--finite-model-find`/`--fmf-bound` typically will slightly degrade CVC4' ability to answer "unsat".

The reason why these techniques are not enabled by default is because some users prefer getting "unknown" to cvc4 timing out.
